File: packages/sdkwork-utils-php/src/CompareUtils.php
<?php

declare(strict_types=1);

namespace Sdkwork\Utils;

final class CompareUtils
{
    public static function deepEqual(mixed $left, mixed $right): bool
    {
        if ($left === $right) {
            return true;
        }
        if ($left === null || $right === null) {
            return false;
        }
        if (is_array($left) && is_array($right) && self::isList($left) && self::isList($right)) {
            if (count($left) !== count($right)) {
                return false;
            }
            foreach ($left as $index => $value) {
                if (!self::deepEqual($value, $right[$index])) {
                    return false;
                }
            }
            return true;
        }
        if (is_array($left) && is_array($right) && !self::isList($left) && !self::isList($right)) {
            if (count($left) !== count($right)) {
                return false;
            }
            foreach ($left as $key => $value) {
                if (!array_key_exists($key, $right) || !self::deepEqual($value, $right[$key])) {
                    return false;
                }
            }
            return true;
        }
        return $left == $right;
    }

    public static function deepClone(mixed $value): mixed
    {
        if (!is_array($value)) {
            return $value;
        }
        $cloned = [];
        foreach ($value as $key => $nested) {
            $cloned[$key] = self::deepClone($nested);
        }
        return $cloned;
    }

    private static function isList(array $value): bool
    {
        return $value === [] || array_keys($value) === range(0, count($value) - 1);
    }
}

File: packages/sdkwork-utils-kotlin/src/main/kotlin/com/sdkwork/utils/CompareUtils.kt
package com.sdkwork.utils

object CompareUtils {
    fun deepEqual(left: Any?, right: Any?): Boolean {
        if (left === right) {
            return true
        }
        if (left == null || right == null) {
            return false
        }
        if (left is Map<*, *> && right is Map<*, *>) {
            if (left.size != right.size) {
                return false
            }
            for ((key, value) in left) {
                if (!right.containsKey(key) || !deepEqual(value, right[key])) {
                    return false
                }
            }
            return true
        }
        if (left is List<*> && right is List<*>) {
            if (left.size != right.size) {
                return false
            }
            for (index in left.indices) {
                if (!deepEqual(left[index], right[index])) {
                    return false
                }
            }
            return true
        }
        return left == right
    }

    fun deepClone(value: Any?): Any? {
        return when (value) {
            is Map<*, *> -> value.entries.associate { (key, nested) -> key to deepClone(nested) }
            is List<*> -> value.map { deepClone(it) }
            else -> value
        }
    }
}
